tlslite/utils/codec.py

```diff
@@ -3,31 +3,195 @@
 
 """Classes for reading/writing binary data (such as TLS records)."""
 
-from .compat import *
+import sys
+import struct
+from struct import pack
+
 
 class Writer(object):
+    """Serialisation helper for complex byte-based structures."""
+
     def __init__(self):
+        """Initialise the serializer with no data."""
         self.bytes = bytearray(0)
 
-    def add(self, x, length):
-        self.bytes += bytearray(length)
-        newIndex = len(self.bytes) - 1
-        for count in range(length):
-            self.bytes[newIndex] = x & 0xFF
-            x >>= 8
-            newIndex -= 1
-        if x != 0:
-            raise ValueError("Can't represent value in specified length")
+    def addOne(self, val):
+        """Add a single-byte wide element to buffer, see add()."""
+        self.bytes.append(val)
+
+    if sys.version_info < (2, 7):
```

Comment:
yes, I know that this is a lot of relatively complex code and lot of juggling with definitions during module load time, but it really is a huge speedup

I have a tool that looks for servers that can't process big Client Hello messages (in the order of 16KiB), to do that is sends about 10 messages, totalling about 128KiB of Client Hello data, with old code, it required just under 4s of CPU time just to generate this stuff, after those changes it went to about 700ms (200ms of which is loading of the rijndael module)

for fuzzer this also will sit in the tight inner loop so I'm afraid that yes, this is worth it

writing our own pack method is not an option, as I want the code to be pure python